packages/context/src/value-promise.ts

```diff
@@ -0,0 +1,160 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+/**
+ * This module contains types for values and/or promises as well as a set of
+ * utility methods to handle values and/or promises.
+ */
+
+// tslint:disable-next-line:no-any
+export type BoundValue = any;
+
+/**
+ * Representing a value or promise. This type is used to represent results of
+ * synchronous/asynchronous resolution of values.
+ */
+export type ValueOrPromise<T> = T | Promise<T>;
+
+export type MapObject<T> = {[name: string]: T};
+
+/**
+ * Check whether a value is a Promise-like instance.
+ * Recognizes both native promises and third-party promise libraries.
+ *
+ * @param value The value to check.
+ */
+export function isPromise<T>(
+  value: T | PromiseLike<T>,
+): value is PromiseLike<T> {
+  if (!value) return false;
+  if (typeof value !== 'object' && typeof value !== 'function') return false;
+  return typeof (value as PromiseLike<T>).then === 'function';
+}
+
+/**
+ * Get nested properties by path
+ * @param value Value of an object
+ * @param path Path to the property
+ */
+export function getDeepProperty(value: BoundValue, path: string) {
+  const props = path.split('.').filter(Boolean);
+  for (const p of props) {
+    if (value == null) {
+      return value;
+    }
+    value = value[p];
+  }
+  return value;
+}
+
+/**
+ * Resolve entries of an object into a new object with the same keys or a
+ * promise of the new object. Please note some of the entries can be resolved
+ * synchronously while others need to be resolved asynchronously.
+ * @param map The original object containing the source entries
+ * @param resolver A function resolves an entry to a value or promise. It will
+ * be invoked with the property value, the property name, and the source object.
+ */
+export function resolveMap<T, V>(
+  map: MapObject<T>,
+  resolver: (v: T, p: string, map: MapObject<T>) => ValueOrPromise<V>,
```

Comment:
Please use `key` instead of `p` here, `p` is too cryptic. I understand it probably means "property", but that does not fit well with the concept of Maps, since they have keys instead of properties.